src/platform/silabs/tracing/SilabsTracing.cpp

```diff
@@ -0,0 +1,164 @@
+#include "SilabsTracing.h"
+#include <cstring>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/PersistentData.h>
+
+#if !CONFIG_BUILD_FOR_HOST_UNIT_TEST
+#include <silabs_utils.h> // for isLogInitialized
+#endif
+
+// The following functions needs to be implemented by the application to allows logging or storing the traces /
+// metrics. If the application does not implement them, the traces will simply be ignored.
+bool __attribute__((weak)) isLogInitialized()
+{
+    return false;
+}
+
+namespace chip {
+namespace Tracing {
+namespace Silabs {
+
+static constexpr size_t kPersistentTimeTrackerBufferMax = SERIALIZED_TIME_TRACKERS_SIZE_BYTES;
+
+struct TimeTrackerStorage : public TimeTracker, PersistentData<kPersistentTimeTrackerBufferMax>
+{
+    // TODO implement the Persistent Array class and use it here for logging the watermark array
+};
+
+SilabsTracer SilabsTracer::sInstance;
+
+SilabsTracer::SilabsTracer()
+{
+    Init();
+}
+
+void SilabsTracer::Init()
+{
+    // Initialize the trace buffer and time trackers
+    memset(mTraceBuffer, 0, sizeof(mTraceBuffer));
+    memset(mWatermark, 0, sizeof(mWatermark));
+}
+
+CHIP_ERROR SilabsTracer::StartTraceStorage(PersistentStorageDelegate * storage)
+{
+    VerifyOrReturnError(nullptr != storage, CHIP_ERROR_INCORRECT_STATE);
+    mStorage = storage;
+    return CHIP_NO_ERROR;
+}
+
+void SilabsTracer::TimeTraceBegin(TimeTraceOperation aOperation)
+{
+    // Log the start time of the operation
+    auto & tracker = mWatermark[static_cast<size_t>(aOperation)];
+    tracker.mTotalCount++;
+    tracker.mStartTime = System::SystemClock().GetMonotonicTimestamp();
+}
+
+void SilabsTracer::TimeTraceEnd(TimeTraceOperation aOperation, CHIP_ERROR error)
+{
+    if (error != CHIP_NO_ERROR)
+    {
+        // Log the error, no need to update the watermark for now as we can get the error count from the
+        // total count - successfull count. We might want to have a separate count for errors in the future.
+        TimeTraceInstant(aOperation);
+        return;
+    }
+
+    // Calculate the duration and update the time tracker
+    auto & tracker   = mWatermark[static_cast<size_t>(aOperation)];
+    tracker.mEndTime = System::SystemClock().GetMonotonicTimestamp();
+    auto duration    = (tracker.mEndTime - tracker.mStartTime).count();
```

Comment:
doesn't System::Clock::Milliseconds64 have additions and comparison operators? We could save a few .count()